pkg/pillar/cmd/zedrouter/acl.go

```diff
@@ -103,9 +105,48 @@ func freeACEId(candidate int32) {
 	}
 }
 
+// Returns name to use for ipset matching a given domain name
+// (ACE match of type "host").
+// Needs to ensure that the ipset name doesn't exceed the length
+// limit of 31 characters imposed by netfilter.
+func hostIpsetName(hostname string) string {
+	maxLen := ipsetNameLenLimit - 5 // leave 5 characters for "ipvX."
+	if len(hostname) < maxLen {
+		return hostname
+	}
+	const (
+		// Minimum number of characters the hash should contain
+		// to bring the probability of collision down to an acceptable level.
+		hashMinLen = 8
+		// Separator between a hostname suffix and a hash-generated prefix.
+		sep = "#"
+	)
+	// Function tries to keep some suffix from the original host name
+	// (to keep the ipset name human-readable and traceable to its source)
+	// and replaces only the remaining subdomains with a hash value.
+	domains := strings.Split(hostname, ".")
+	var suffixLen, i int
+	for i = len(domains); i > 0; i-- {
+		if suffixLen+len(domains[i-1])+1+hashMinLen > maxLen {
+			break
+		}
+		suffixLen += len(domains[i-1]) + 1
+	}
+	domains = domains[i:]
+	suffix := strings.Join(domains, ".")
```

Comment:
Can't tell whether this correctly handles a name with a single label > 26 characters.
Normally the DNS TLDs are short, but an enterprise DNS might resolve local names (such as shop.verylongcompanynamewhichmakesnosense) hence they might be used in ACLs.

Another case to check is whether it handles the case of a trailing dot (some folks might use "google.com." which is technically more correct than "google.com")

Reply:
Added unit test to cover all these cases.
For over-the-limit TLD there will be no preserved suffix, just the generated hash. The `#` separator is still included (at the very end of the ipset name), so that it is clear that this is a hash and not an actual domain.
Please look at `acl_test.go` to see how the generated name looks like for every corner case and feel free to leave suggestions for any further improvements.